xds/googledirectpath/googlec2p.go

```
@@ -0,0 +1,219 @@
+	}
+
+	ipv6EnabledMetadata = &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			ipv6CapableMetadataName: {
+				Kind: &structpb.Value_BoolValue{BoolValue: true},
+			},
+		},
+	}
+)
+
+func init() {
+	resolver.Register(&c2pResolverBuilder{})
+}
+
+type c2pResolverBuilder struct{}
+
+// Build helps implement the resolver.Builder interface.
+func (b *c2pResolverBuilder) Build(t resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	ret := &c2pResolver{
+		target: t,
+		cc:     cc,
+		opts:   opts,
+	}
+	// start does I/O. Run it in a goroutine.
+	go ret.start()
+	return ret, nil
+}
+
+// Name helps implement the resolver.Builder interface.
+func (*c2pResolverBuilder) Scheme() string {
+	return c2pScheme
+}
+
+type c2pResolver struct {
+	target resolver.Target
+	cc     resolver.ClientConn
+	opts   resolver.BuildOptions
+
+	mu    sync.Mutex
+	done  bool
+	child resolver.Resolver
+}
+
+func (r *c2pResolver) start() {
+	if !runDirectPath() {
+		r.startChild("dns")
+		return
+	}
+
+	nodeCopy := proto.Clone(defaultNode).(*v3corepb.Node)
+	nodeCopy.Locality = &v3corepb.Locality{Zone: getZone()}
+	if getIPv6Capable() {
+		nodeCopy.Metadata = ipv6EnabledMetadata
+	}
+	config := &bootstrap.Config{
+		BalancerName: tdURL,
+		Creds:        grpc.WithCredentialsBundle(google.NewDefaultCredentials()),
+		TransportAPI: version.TransportV3,
+		NodeProto:    nodeCopy,
+	}
+
+	// Create singleton xds client with this config. The xds client will be
+	// used by the xds resolver later.
+	err := newClientWithConfig(config)
+	if err != nil {
+		r.cc.ReportError(fmt.Errorf("failed to start xDS client: %v", err))
+		return
+	}
+
+	r.startChild("xds")
+}
+
+func (r *c2pResolver) startChild(scheme string) {
+	t := r.target
+	t.Scheme = scheme
+
+	b, ok := childBuilders[scheme]
+	if !ok {
+		logger.Errorf("unknown child scheme: %q", scheme)
+		r.cc.ReportError(fmt.Errorf("unknown child scheme in c2p resolver: %q", scheme))
+		return
+	}
```

Pass the builder instead so this can't happen here?

---

Moved `Get` to build, and fail fast.